tests/shared/src/test/scala/cats/tests/FunctorSuite.scala

```diff
@@ -45,6 +45,18 @@ class FunctorSuite extends CatsSuite {
     }
   }
 
+  test("mapOrKeep maps matching elements preserving structure") {
+    forAll { (l: List[Int], o: Option[Int], m: Map[String, Int]) =>
+      assert(l.mapOrKeep { case i if i % 2 == 0 => i + 1 } === l.map(i => if (i % 2 == 0) i + 1 else i))
+      assert(
+        o.mapOrKeep { case i if i > 0 => i + 1 } === (if (o.nonEmpty) if (o.get > 0) Some(o.get + 1) else o else None)
+      )
+      assert(
+        m.mapOrKeep { case v if v % 2 == 0 => v + 1 } === m.map { case (k, v) => k -> (if (v % 2 == 0) v + 1 else v) }
+      )
+    }
```

Comment:
Although it is a valid test case - no doubt in that, but it is not really necessary to invent a custom test logic there:
```scala
    forAll { (l: List[Int], o: Option[Int], m: Map[String, Int], pf: PartialFunction[Int, Number]) =>
      assertEquals(l.mapOrKeep(pf), l.map(x => pf.applyOrElse(x, _ => x: Integer))
      // same for `o: Option[Int]` and `m: Map[String, Int]`
    }
```
It would be way clearer and more generic.
Also:
1. `PartialFunction[Int, Number]` so it check that the `A1 >: A` constraint holds too.
2. I would probably go with `assertEquals` instead of just `assert` because the former provides better reporting in a case of test failures.

That said, I feel that this equivalence could be just a part of the `FunctorLaws`:
```
Functor[A].mapOrKeep(a)(pf) <-> Functor[A].map(a => pf.applyOrElse(a, _ => a: A1)
```
If we could add such a rule to the laws, we would not need to check for special cases individually – all possible implementations would be got checked automatically.

Reply:
@satorg i guess I'm missing something again, but `Int` is not a subtype of `Number`
`type arguments [Int,Number] do not conform to method m's type parameter bounds [A,A1 >: A]`

Reply:
Correct, `Int` is definitely not, but `Integer` is.

Reply:
But there are no generators for Integer. As I mentioned in another comment, I didn't find any examples of A and A1 with A1 >: A

Reply:
Well, I thought that since there's a generator for `Int` then it wouldn't be too difficult to get one for `Integer` via boxing.

But on the second thought, maybe it's not really worth the hassle – perhaps `PartialFunction[Int, Int]` should be just enough for this sort of test.

Besides, if some method needs tests for syntax correctness (which may also include type correspondence/variance testing), then the `SyntaxSuite` could be better bet for that.
https://github.com/typelevel/cats/blob/47fbad764ad8026d4e9f8999c7f6233c666f0df8/tests/shared/src/test/scala/cats/tests/SyntaxSuite.scala#L339

Reply:
given that we agreed to use PF[A, A] in the law test, seems like we should be covered by it, and it also uses syntax